File: src/entity.py
from level import Level

class Entity():
    def __init__(self, size, position, mass, drag) -> None:
        self.size = size
        self.mass = mass

        self.position = position
        self.velocity = [0, 0]
        self.acceleration = [0, 0]

        self.drag = drag

        self.level: Level
        self.level = None

        self.isAffectedByGravity = True


    def set_level(self, level: Level) -> None:
        self.level = level


    def setIsAffectedByGravity(self, isAffectedByGravity: bool) -> None:
        self.isAffectedByGravity = isAffectedByGravity


    def __gravity(self, delta_time) -> None:
        if self.isAffectedByGravity and self.level:
            self.velocity[1] += self.level.gravity * delta_time * 60


    def move(self, delta_time) -> tuple:
        self.__gravity(delta_time)

        drag_froce = (-self.velocity[0] * self.drag, -self.velocity[1] * self.drag)

        self.velocity[0] += (self.acceleration[0] + drag_froce[0]) * delta_time * 60
        self.velocity[1] += (self.acceleration[1] + drag_froce[1]) * delta_time * 60

        if -0.01 <= self.velocity[0] <= 0.01:
            self.velocity[0] = 0

        # self.position[0] += self.velocity[0] * delta_time * 60
        # self.position[1] += self.velocity[1] * delta_time * 60

        new_position = self.position[:]
        new_position[0] += self.velocity[0] * delta_time * 60
        new_position[1] += self.velocity[1] * delta_time * 60

        has_moved = new_position[0]-self.position[0] != 0 or new_position[1]-self.position[1] != 0

        if has_moved:
            solve = self.level.check_for_collisions_on_path(self, self.position, new_position)
            if solve != None:
                hitbox = solve[0]
                collision_line = solve[1]
                collision_with_horizontal = collision_line[1][1] == 0

                if collision_with_horizontal:
                    self.position[0] = new_position[0]
                    self.velocity[1] = 0
                    self.acceleration[1] = 0

                else:
                    self.position[1] = new_position[1]
                    self.velocity[0] = 0
                    self.acceleration[0] = 0

            else:
                self.position = new_position

            
        # print(self.position)

        return self.position

File: src/entity.c
#include "entity.h"
#include "level.h"

#include "raymath.h"
#include "stdlib.h"
#include "cJSON.h"
#include <sys/stat.h>
#include <sys/types.h>
#include "stdlib.h"
#include <stdio.h>
#include "string.h"

Entity* CreateEntity(Vector2 position, char* tag, bool isAffectedByGravity, float mass, Vector2 size, float drag, int uuid) {
    Entity* entity = malloc(sizeof(Entity));

    char *path = "resources/data/entity_tags.json";
    FILE *json_file_pointer = fopen(path, "r");
    
    if (json_file_pointer == NULL)
    {
        // laod error handling
    }

    struct stat fileStats; 

    if (stat(path, &fileStats) != 0)
    {
        // file stat error handling
    }

    int fileSize = fileStats.st_size;
    char dataBuffer[fileSize];
    fread(dataBuffer, 1, sizeof(dataBuffer), json_file_pointer); 
    
    fclose(json_file_pointer);


    cJSON *tags_json = cJSON_Parse(dataBuffer);

    cJSON *tag_list = cJSON_GetObjectItem(tags_json, tag);
    int numberOfTags = cJSON_GetArraySize(tag_list);
    char *tags = malloc(numberOfTags*20);
    strcpy(tags, "");
    
    for (int i=0; i<numberOfTags; i++)
    {
        strcat(tags, "_");
        strcat(tags, cJSON_GetArrayItem(tag_list, i)->valuestring);
    }

    cJSON_Delete(tags_json);

    entity[0] = (Entity)
    {
        .position = position,
        .isAffectedByGravity = isAffectedByGravity,
        .mass = mass,
        .size = size, 
        .drag = drag,

        .velocity = (Vector2){0, 0},
        .acceleration = (Vector2) {0, 0},
        .onGround = false,


        .hitbox = (Hitbox)
        {
            .rect = (Rectangle)
            {
                .x = position.x - size.x / 2,
                .y = position.y - size.y / 2,
                .width = size.x,
                .height = size.y
            },
            .tag = tag
        },

        .uuid = uuid,
        .tags = tags,
        .state = idle
    };
    
    return entity;
}

void UpdateEntity(Entity* entity, float gravity, float deltaTime)
{
    if(!entity->onGround && entity->isAffectedByGravity)
    {
        entity->velocity.y += gravity * deltaTime * 60;
    }

    entity->velocity = Vector2Add(entity->velocity, Vector2Scale(entity->acceleration, deltaTime * 60));
    
    entity->velocity.x -= entity->velocity.x * entity->drag;

    entity->position = Vector2Add(entity->position, Vector2Scale(entity->velocity, deltaTime * 60));

    entity->hitbox.rect = (Rectangle) {
        .x = entity->position.x - entity->size.x/2,
        .y = entity->position.y - entity->size.y/2,
        .width = entity->size.x,
        .height = entity->size.y
    };
}

void DrawEntity(Entity entity)
{
    DrawRectangle(entity.position.x - (entity.size.x / 2), entity.position.y - (entity.size.y / 2), entity.size.x, entity.size.y, RED);
}

bool EntityHasTag(Entity* entity, char* tag)
{
    char* entityTags = malloc(sizeof(entity->tags));
    strcpy(entityTags, entity->tags);

    char* token = strtok(entity->tags, "_");

    while (token != NULL)
    {
        if (!strcmp(token, tag))
        {
            strcpy(entity->tags, entityTags);
            return true;
        }
        token = strtok(NULL, "_");
    }

    strcpy(entity->tags, entityTags);
    return false;
}

